### PR TITLE
bind: remove hard-coded `allow-query` config setting

### PR DESCRIPTION
**Pull request by mrobbetts**
###### Description of changes

The `bind` module currently hard-codes the value of `allow-query` for every zone, making it not configurable (if you add it to `extraConfig`, it is duplicated in the output config file, and you get an error).

I cannot see a reason why `allow-query` is particularly special among the multitude of config parameters (and, according to [this doc](https://www.zytrax.com/books/dns/ch7/queries.html#allow-query), it defaults to `any`, which is just what we have here anyway), and it seems the most idiomatic way to control it – with the module in its current form – is to use `extraConfig`.

I have this up and running with an entry in `extraConfig` on my setup, and it is working perfectly.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by peti — APPROVED** [review]
Sounds fine to me, especially if this is the default anyway.

**Review by teto — APPROVED** [review]

**Comment by mrobbetts** [comment]
Thanks, @teto, @peti!

Now, how do we get this merged? I don't have write access...

**Comment by Ma27** [comment]
Technically this is a breaking change (I relied on this in a testing environment without knowing of this detail) and only found out because I looked through the git history, so I think this should have an entry in the release notes.

cc @NixOS/nixos-release-managers

**Comment by peti** [comment]
> Technically this is a breaking change.

Why do you think it is? The PR doesn't actually change anything. It just opens the door for users to make a change by specifying allow-query themselves.

**Comment by Ma27** [comment]
Well, if you're relying on the behavior on 22.11 - that you can query named from "anywhere" (perhaps without knowing that) - then this will break *silently* when upgrading to 23.05. The only ways to find out are to either go through the git history - like I did - or to diff the configs from before/after the upgrade.

None of that is very nice. If we have such a subtle - yet noticeable - change in behavior, I think it's good to have this documented in the release notes.

**Comment by peti** [comment]
I am sorry, but I don't follow. This PR simply removed a configuration option that repeated BIND's default value. In my understanding, this PR does not change anything. The setting of allow-setting was completely redundant.

What am I missing?

**Comment by Ma27** [comment]
Perhaps it helps to show the problem I actually had: https://github.com/NixOS/nixpkgs/pull/140669/commits/4f4c56d2f13c8b2ecb07d48d2d4552f7542412c5

After merging in `master` the test didn't work anymore and this commit fixed the test. I relied on the behavior with this option being configured and because it got removed from this module[1] I had to set it on my own to restore the behavior I was relying on.

[1] To avoid misunderstandings: I think this change is perfectly reasonable, but to avoid surprises for others with the same problem I think it shold be documented.

**Comment by riotbib** [comment]
Hi.

> None of that is very nice. If we have such a subtle - yet noticeable - change in behavior, I think it's good to have this documented in the release notes.

@Ma27, I think so too – as far as I understand the issue. May I ask to draft a notice for the [Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md)? (Asking as the Release Editor for 23.05.) Otherwise I'd give it a try in the following days, but your input is greatly appreciated.

Thanks for your work and input.

**Comment by Ma27** [comment]
@peti the allow-query was implicitly added by the module until this change. Now that it's removed I had to add it by myself to keep the test working.

@riotbib sure, will try to do that soonish :)

**Comment by peti** [comment]
> @peti the allow-query was implicitly added by the module until this change. Now that it's removed I had to add it by myself to keep the test working.
> 
> @riotbib sure, will try to do that soonish :)

Maybe saying it for the 3rd time is a charm: `allow-query { any }` is the DEFAULT SETTING. Specifying that directive is a NO-OP. It does not actually do anything. 

So, if you're test requires this option to be specified, then this is totally unexpected behavior and I would like to understand WHY specifying that option makes a difference.

**Comment by RaitoBezarius** [comment]
> > @peti the allow-query was implicitly added by the module until this change. Now that it's removed I had to add it by myself to keep the test working.
> > @riotbib sure, will try to do that soonish :)
> 
> Maybe saying it for the 3rd time is a charm: `allow-query { any }` is the DEFAULT SETTING. Specifying that directive is a NO-OP. It does not actually do anything.
> 
> So, if your test requires this option to be specified, then this is totally unexpected behavior and I would like to understand WHY specifying that option makes a difference.

Not a BIND expert but, wouldn't this default depend on which interface/IP do you listen on? See https://github.com/NixOS/nixpkgs/blob/4f4c56d2f13c8b2ecb07d48d2d4552f7542412c5/nixos/tests/containers-next/basic.nix#L40-L61 for the actual code triggering the issue, it can be run locally and tested.

**Comment by Ma27** [comment]
AFAIU if `allow-query` isn't specified per-zone (change from this PR), then the global `allow-query` is used if used. The module defines an ACL called `cachenetworks` (which defaults to `127.0.0.0/24`) and that is globally used for `allow-query`. That means that there's in fact a behavioral change because `any` isn't the case anymore.

**Comment by mrobbetts** [comment]
Argh, sorry everyone, I didn't see that the global `allow-query` is being set in the module too 🤦 This makes total sense.

And then, from the help text of the `cachenetworks` option:

```
What networks are allowed to use us as a resolver.  Note
that this is for recursive queries -- all networks are
allowed to query zones configured with the `zones` option.
It is recommended that you limit cacheNetworks to avoid your
server being used for DNS amplification attacks.
```

it sounds like there is some BIND subtlety here (I am not a BIND expert whatsoever). Are recursive queries allowed according to the global definition of `allow-query`, but per-zone queries limited separately, to the zone scope's definition?

In that case, I expect this text needs to be reworked, also, right?

**Comment by mrobbetts** [comment]
I wonder if a good solution is to add `allow-query` to the `bindZoneOptions` and have it default to `any`.

But then, would we also need to add all other options at this level, like `allow-update`, which would be bloaty...

**Comment by peti** [comment]
> I wonder if a good solution is to add `allow-query` to the `bindZoneOptions` and have it default to `any`.

Yes, that seems like the only way to allow users to change that setting without changing the behavior of servers that don't specify the option.

**Comment by mrobbetts** [comment]
@peti, would it be satisfactory to you to add _only_ `allow-query` to `bindZoneOptions`, while making it clear (with a comment or so) that it only exists there for backwards compatibility? Since, it doesn't really make sense for it to be there alone on its own merits, but we don't want to add all of its ~30 siblings.

(And, just to ask one last time: is it a hard pass from you on changing the behavior and just documenting it appropriately?)

**Comment by mrobbetts** [comment]
@peti @Ma27 @riotbib @yu-re-ka

I took a stab at a better version (#224776). Please let me know what you think.

**Pull request merged**
